ext/OptimizationEnzymeExt.jl

```diff
@@ -15,13 +17,57 @@ isdefined(Base, :get_extension) ? (using Enzyme) : (using ..Enzyme)
     end
 end
 
+function inner_grad(θ, bθ, f, p, args::Vararg{Any, N}) where N
+    Enzyme.autodiff_deferred(Enzyme.Reverse,
+        Const(firstapply),
+        Active,
+        Const(f),
+        Enzyme.Duplicated(θ, bθ),
+        Const(p),
+        Const.(args)...),
+    return nothing
+end
+
+function hv_f2_alloc(x, f, p, args...)
+    dx = Enzyme.make_zero(x)
+    Enzyme.autodiff_deferred(Enzyme.Reverse,
+        firstapply,
+        Active,
+        f,
+        Enzyme.Duplicated(x, dx),
+        Const(p),
+        Const.(args)...)
+    return dx
+end
+
+function inner_cons(x, fcons::Function, p::Union{SciMLBase.NullParameters, Nothing}, num_cons::Int, i::Int)
+    res = zeros(eltype(x), num_cons)
+    fcons(res, x, p)
+    return res[i]
+end
+
+function cons_f2(x, dx, fcons, p, num_cons, i)
+    Enzyme.autodiff_deferred(Enzyme.Reverse, inner_cons, Active, Enzyme.Duplicated(x, dx), Const(fcons), Const(p), Const(num_cons), Const(i))
+    return nothing
+end
+
+function inner_cons_oop(x::Vector{Float64}, fcons::Function, p::Union{SciMLBase.NullParameters, Nothing}, i::Int)::Float64
+    return fcons(x, p)[i]
+end
+
+function cons_f2_oop(x, dx, fcons, p, i)
+    Enzyme.autodiff_deferred(Enzyme.Reverse, inner_cons_oop, Active, Enzyme.Duplicated(x, dx), Const(fcons), Const(p), Const(i))
```

Comment:
This has trouble being inferred (I made the `inner_cons_oop` strongly typed to try to force it but still) when there is more than one constraint but it works for the one constraint test

cc: @wmoses incase it seems obvious why that might happen

Reply:
https://github.com/SciML/OptimizationBase.jl/actions/runs/9522957627/job/26253527285?pr=60

Reply:
Retry on latest enzyme, there's been some fixes and I'm curious if that resolves things for you

Reply:
cc @ChrisRackauckas

Reply:
Looks like it works!

Reply:
Awesome! You should try to also move the other closures, etc outside as well. It may help improve compatibility/performance, but more critically it means that things don't need to be recompiled each time [since each time a closure is defined, its a new function and the derivative won't be cached.